draft-ietf-webtrans-http3.md

```diff
@@ -218,6 +218,31 @@ session have alternative mechanisms:
   limiting {{!RFC6585}}.  Unlike the previous method, this signal is directly
   propagated to the application.
 
+## Prioritization
+
+WebTransport does not provide any priority signalling mechanism for streams and
+datagrams within a WebTransport session; such mechanisms can be defined by
+application protocols using WebTransport.
+
+The Priority header {{!RFC9218}} MAY be set in a WebTransport CONNECT request or
+response; clients MAY also reprioritize WebTransport sessions by sending
+PRIORITY_UPDATE frames ({{Section 7 of RFC9218}}).  These priority signals
+SHOULD be included by the recipient as input to the scheduling process for all
+data it sends in the enclosing WebTransport session, including Capsules,
+WebTransport streams and datagrams.
+
```

Comment:
It's not just the recipient, consider the case where a client sends signal X and something coordinating with the server says Y, the server will likely pick Y due to https://www.rfc-editor.org/rfc/rfc9218.html#section-8

That's what's not clear here. In HTTP the interaction expectations were set out clearly. WebTransport is a bit different and we should make sure everyone understands that. 

So as a concrete strawman, I'd phrase it something like this (and run into your second para example, which is fine as it is.

```suggestion
WebTransport sessions are initiated using the CONNECT method. While  {{Section
11 of !RFC9218}} describes how extensible priorities can be applied to CONNECT,
WebTransport extends the types of data that are exchanged in relation to the
request and response, which requires additional considerations. WebTransport
CONNECT requests and responses MAY contain the Priority header field  ({{Section
5 of RFC9218}}); clients MAY reprioritize by sending PRIORITY_UPDATE frames
({{Section 7 of RFC9218}}). In extension to {{RFC9218}}, it is RECOMMENDED that
the scheduling guidance in both {{Section 9 of RFC9218}} and {{Section 10 of
RFC9218}} is applied by clients and servers for all data that they send in the
enclosing WebTransport session, including Capsules, WebTransport streams and
datagrams. Note that the merging considerations given in {{Section 8 of
RFC9218}} also apply. For example, a client that receives a response Priority
header field could alter its view of a WebTransport session priority and alter
the scheduling of outgoing data as a result. 

Endpoints that prioritize WebTransport sessions need to consider how they
interact with other sessions or requests on the same HTTP/3 connection.
```

Reply:
Thanks, overall, your text is much better. 

> it is RECOMMENDED that the scheduling guidance in both {{Section 9 of RFC9218}} and {{Section 10 of RFC9218}} is applied by clients and servers for all data that they send in the enclosing WebTransport session, including Capsules, WebTransport streams and datagrams

I think the bit I want to clarify is that prioritizing WebTransport streams or datagrams (which can be defined elsewhere) is only intended to prioritize them against other data in the same WebTransport session, and the Priority header is what prioritizes the session itself. 

Admittedly, my text didn't convey this concept clearly either.

Reply:
ah got it. I agree on the concept. I think we can highlight that with a little bit more iteration of the text in this section. 

The key point that does need to be covered, if I understand your intent for whole-session priotization,  is that RFC 9218 Section 9 says client MAY apply signals, and in WebTransport you want to upgrade it to a SHOULD.

Reply:
Oh actually the key point was the 'dependent streams and datagrams get prioritized within the session and not across sessions'; asking the client to respect the server's preference on send with a SHOULD was more of a target of opportunity, and I could be convinced to leave it as a MAY.  I can contrive a use case where there are pooled sessions and the server wants to elevate the priority of one, or indicate a pooled POST is lower priority than some WT session, but a) I'm not sure how real they are and b) I don't know if the difference between MAY and SHOULD makes any difference.

Reply:
The concepts are tangential to me. I.e.

1. A client SHOULD employ some form of prioritization when pooling (or else you might do something naive and cause starvation)
2. A client SHOULD apply the scheduling guidance in section 9 when sending session data (a deviation from section 8, which was written to be interpreted very openly)
3. A client can send a signal to indicate it's send priority to the origin (or intermediary)
4. A server might want to consider the client priority signal but it doesn't have to
5. A client might want to consider the server priority but it doesn't have to

Reply:
I would drop 4 & 5.